File: ClaudeUsageMonitor/Services/NotificationService.swift
import Foundation
import UserNotifications

final class NotificationService {
    static let shared = NotificationService()
    private init() {}

    /// Thresholds already notified in the current session window.
    private var notifiedThresholds = Set<Int>()

    /// Last known reset date — used to detect when a new window starts.
    private var lastKnownResetDate: Date?

    // MARK: - Permission

    func requestPermission() {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { _, _ in }
    }

    // MARK: - Main entry point

    func checkAndNotify(data: UsageData) {
        checkUsageThresholds(data: data)
        checkSessionReset(data: data)
    }

    // MARK: - Usage threshold alerts

    private func checkUsageThresholds(data: UsageData) {
        // Use sessionPercentage (rate-limit window), fall back to usagePercentage
        let pct = data.hasSessionData ? data.sessionPercentage : data.usagePercentage

        let config: [(threshold: Int, title: String, body: String)] = [
            (75, "Halfway through your session",
                 "Consider wrapping up long threads. Start fresh conversations for new topics."),
            (80, "Session at 80%",
                 "Avoid new long projects or file uploads. Best for: quick questions, short edits, code review."),
            (90, "Session at 90% — act fast",
                 "~10% left. Finish your current task and save important outputs before the limit hits."),
            (95, "Almost out",
                 "Save your work now. \(data.sessionResetLabel ?? "Session resets soon")."),
            (100, "Claude Limit Reached",
                  "You've used your full quota. \(data.sessionResetLabel ?? "Resets soon").")
        ]

        for item in config {
            let fraction = Double(item.threshold) / 100.0
            if pct >= fraction {
                guard !notifiedThresholds.contains(item.threshold) else { continue }
                notifiedThresholds.insert(item.threshold)
                sendAlert(title: item.title, body: item.body, id: "claude-tip-\(item.threshold)")
            } else {
                notifiedThresholds.remove(item.threshold)
            }
        }
    }

    private func sendAlert(title: String, body: String, id: String) {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body  = body
        content.sound = .default
        let request = UNNotificationRequest(identifier: id, content: content, trigger: nil)
        UNUserNotificationCenter.current().add(request, withCompletionHandler: nil)
    }

    // MARK: - Session reset detection

    private func checkSessionReset(data: UsageData) {
        guard let newReset = data.resetDate else { return }

        if let known = lastKnownResetDate, newReset > known.addingTimeInterval(3600) {
            notifiedThresholds.removeAll()
            sendAlert(
                title: "Claude Session Reset",
                body: "Your usage window has reset. You have a full quota available.",
                id: "claude-reset-\(Int(Date().timeIntervalSince1970))"
            )
        }

        lastKnownResetDate = newReset
    }
}

File: windows/notification_service.py
"""Windows toast notification service for Claude Code usage alerts."""

from datetime import datetime
from typing import Optional

from data_reader import RateLimitData

_THRESHOLDS = [
    (75, "Halfway through your session",
         "Consider wrapping up long threads. Start fresh conversations for new topics."),
    (80, "Session at 80%",
         "Avoid new long projects or file uploads. Best for: quick questions, short edits, code review."),
    (90, "Session at 90% — act fast",
         "~10% left. Finish your current task and save important outputs before the limit hits."),
    (95, "Almost out",
         "Save your work now."),
    (100, "Claude Limit Reached",
          "You've used your full quota."),
]


def _send_toast(title: str, body: str) -> None:
    try:
        from winotify import Notification
        Notification(app_id="Claude Usage Monitor", title=title, msg=body).show()
    except Exception as e:
        print(f"[notification] {e}")


class NotificationService:
    """Tracks usage thresholds and fires Windows toast notifications."""

    def __init__(self):
        self._notified: set[int] = set()
        self._last_reset_at: Optional[datetime] = None

    def check_and_notify(self, data: RateLimitData) -> None:
        self._check_thresholds(data)
        self._check_session_reset(data)

    def _check_thresholds(self, data: RateLimitData) -> None:
        window = data.five_hour
        if window is None or window.used_percentage is None:
            return

        pct = window.used_percentage

        for threshold, title, body in _THRESHOLDS:
            if pct >= threshold:
                if threshold not in self._notified:
                    self._notified.add(threshold)
                    if threshold >= 95 and window.resets_in:
                        body = f"{body} Resets in {window.resets_in}."
                    _send_toast(title, body)
            else:
                self._notified.discard(threshold)

    def _check_session_reset(self, data: RateLimitData) -> None:
        window = data.five_hour
        if window is None or window.resets_at is None:
            return

        new_reset = window.resets_at
        if self._last_reset_at is not None:
            if (new_reset - self._last_reset_at).total_seconds() > 3600:
                self._notified.clear()
                _send_toast(
                    "Claude Session Reset",
                    "Your usage window has reset. You have a full quota available.",
                )

        self._last_reset_at = new_reset
